Renault: Provide a meaningful error when not paired

ickeundso opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/3535

ickeundso committed:
#3535 Renault: Provide a meaningful error when not paired

- log.ERROR.Println added to kamereonVehicles
- no logic changes only a log messages is added

It would also be possible to move this to the ==ACTIVE branch. But I don't know how the kamereon API will behave with other versions

ickeundso committed:
Renault: Provide a meaningful error when not paired

- do not only log the not paired vehicle but create an error response the missconfigured / non-paired vehicle

ickeundso committed:
Merge remote-tracking branch 'origin/master' into #3535

ickeundso committed:
Merge remote-tracking branch 'origin/master' into #3535

ickeundso committed:
Renault: Provide a meaningful error when not paired

- do not only log the not paired vehicle but create an error response the missconfigured / non-paired vehicle
- do not check for role "MAIN_DRIVER" string, just check of role existence. The "ConnectedDriver.Role" is not documented at https://renault-api.readthedocs.io/en/latest/reference/kamereon.html#renault_api.kamereon.models.KamereonVehiclesLink but I can reproduce the behaviour. For the case the vehicle is paired the ConnectedDriver.Role attribute is there an has the "MAIN_DRIVER" value and if the vihicle is not paired the ConnectedDriver.Role attribute is not there, the API can be tested of empty string.

ickeundso committed:
Renault: Provide a meaningful error when not paired

- remove not needed local var

ickeundso committed:
Renault: Provide a meaningful error when not paired

- add test for vehicle/renault.go
- use vehicle from account if no one is cofiggured on evcc or use a matching vehicle

ickeundso committed:
add .idea to .gitignore

ickeundso committed:
Merge branch 'master' into #3535

ickeundso committed:
Merge branch 'master' into #3535

andig commented:
@ickeundso Du vermischst wieder die Funktionen. `kamereonVehicles` soll die Fahrzeuge zurück geben, nicht mehr und nicht weniger. Dafür braucht es keine VIN. lass es doch wie's ist (ggf. ein `[]Vehicle` zurück geben statt `[]string`) und gib dem Vehicle dann eine `IsActive` und `IsPaired` methode. Aufruf und Auswertung dann aus `NewRenault`.

andig commented on vehicle/renault.go:
```diff
@@ -308,20 +313,37 @@ func (v *Renault) kamereonPerson(personID string) (string, error) {
 	var vehicles []string
 	if err == nil {
-		for _, v := range res.VehicleLinks {
-			if strings.ToUpper(v.Status) == "ACTIVE" {
```
...oder Du machst es ganz einfach und gibst hier nur aktive *UND* gepairte Fahrzeuge zurück. Dann kann ALLES andere so bleiben wie es ist...

ickeundso replied:
@andig , so richtig verstehe ich Deine Anmerkungen nicht. `kamereonVehicles` gibt auch in der master version ein string[]. Nur ist hier das string[] nicht vollständig gefiltert. Ich habe ja eigentlich genau nur das gemacht was du in dem Kommentar hier oben anmerkst nämlich die Vehicles werden zusätzlich nach  ihrer Rolle gefiltert. So das ich des weiteren ein error logging mit hinzugefügt habe ändert ja nicht an der semantic was die func zuvor zurückgegeben hat.  Der weitere Filter kam ja erst durch Deinen Kommentar beim meinem ersten Versuch hinzu. Und den finde ich auch richtig. evcc ruft diese func ja mehrfach auf für jedes vehicle was in der config hinterlegt ist und hier wird das mit eine liste von vehicles gemerged die einem kamereon account zugeordnet sind. Das Problem warum es hier falsch erscheint, ist meiner Meinung nach dass das NewRenaultFromConfig interface ein string[] verlangt. Denn die Liste der config vehicles wird nicht immer der Liste der account vehicles entsprechen. Eigentlich, durch den mehrfachen Aufruf dieser func (anzahl vehicle in der cvcc config) und die Bindung dort an genau einen account sollte die func nur immer ein vehicle zurückgeben. 

Aber ja ich könne in New.. auch eine andere func die nur die validen vehicles zurückgeben lassen und `kamereonVehicles` liefert alle account vehicles zurück , aber auch die nicht activen sonst würde man wirklich die semantic vermischen (account vehicles und in ecvv verwendbare vehicles. Also auch der bisher verwendete filter nach "ACTIVE" würde dann in die neue func wandern.

andig commented:
Ich machs mal zu, gerne neuer Anlauf wenn Du Lust hast.

ickeundso committed:
Renault: Provide a meaningful error when not paired

- add test for vehicle/renault.go
- use vehicle from account if no one is configured on evcc or use a matching vehicle
- fix linter warn

andig commented:
Ich glaub wir müssen uns nur einigen wie es werden soll: entweder filtern wir alle unzugänglichen Fahrzeuge stillschweigend raus (so ist es heute). Oder wir geben alle Fahrzeuge zurück (es ist ja die vehicles Funktion) und lassen übergeordnet prüfen was davon zu verwenden ist. Auf die Art bleibt der Code sauber, Beides wäre möglich.

andig commented:
Mein Vorschlag wäre https://github.com/evcc-io/evcc/pull/3706 (ich bekomms grad nicht in Deinen PR gepusht, irgendein Git-Problem). Wär das ok für Dich?

andig committed:
simplify

andig commented:
So, jetzt ist es hier drin. Please check.

andig commented:
@ickeundso passt die Funktion? Dann müssten nur noch die Tests entfernt werden oder auf den Typ umgebaut.

ickeundso commented:
HI @andig , haha, da kann ich sehr gut mit leben. Ist ja wesentlich weiter gedacht Deine Lösung. Ja die test könnte ich umbauen, entweder hier oder ich mach dafür nach Deinem master merge einen separaten PR auf und wir schieben die Test nach, das wäre dann zwar nicht ganz so wie man es macht, aber ....

andig commented:
Genau. Dann schmeiss gerne den Test erstmal raus, wäre aber schön wenn Du probieren könntest ob es funktioniert- dafür fehlt mir das Fahrzeug...

andig committed:
wip
